### ISSUE TITLE
User impersonation in C# code - does it work when "Integrated Security=True"?

### ISSUE THREAD

**Issue by cnayan** [dotnet/SqlClient#768]
This is continuation of issue: [https://github.com/dotnet/SqlClient/issues/432](https://github.com/dotnet/SqlClient/issues/432)

**Context**: I am unable to fetch data from DB with `Integrated Security=True`, when I try to impersonate another user (using impersonation code as documented on MSDN https://docs.microsoft.com/en-us/dotnet/api/system.security.principal.windowsidentity.runimpersonated?view=dotnet-plat-ext-3.1)

- There are 2 SQL servers: one which is directly connectable in my domain A, and another which is connectable when I impersonate as user in domain B.
- When I `runas` SSMS or `SQL Server plugin` in VS Code, then I am able to connect to SQL server in domain B, as user of domain B. Connectivity is not an issue here.
- When I try to impersonate as user from domain B in C# code, then SQL connection doesn't happen.

Now....

After spending a few months, I understand that WindowsIdentity.GetCurrent() is not returning the user that the code has been launched as. This is not your code's fault, but because you are using it (here: https://github.com/dotnet/SqlClient/blob/a9dfe480eff5dd344518eab5198654eff969fb6e/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/ProviderBase/DbConnectionPoolIdentity.Windows.cs#L23), this affects how you perceive who is using the SqlConnection.

To test, dump info of identity returned by `WindowsIdentity.GetCurrent()`

```c#
            using (WindowsIdentity identity = WindowsIdentity.GetCurrent())
            {
                SecurityIdentifier user = identity.User;
                bool isNetwork = user.IsWellKnown(WellKnownSidType.NetworkSid);
                string sidString = user.Value;
                Console.WriteLine($"User = {identity.Name}, Is Network = {isNetwork}, SID = {sidString}");
            }
```

This will dump same information when you program is executed from command line as is, and when executed as another user. I used a user from another domain, and executed the code using `runas` tool.

There was no authentication error, just in case you may be wondering.

I am puzzled because SSMS and SQL Server plugin of VS Code are able to connect and run queries under `runas` impersonation. Why is this C# code not able to find the right user being run as?

What am I missing, that above 2 applications have mastered?

Please help.

**Comment by karinazhou** [comment]
Hi @cnayan ,

I tested running a standalone application runas another userA which is different from the userB I login to my computer. The result I can see is that the `identity.Name` still shows the userB even though the application is actually runas userA. 

SSMS uses System.Data.SqlClient internally and probably SQL Server plugin of VS Code does this too. From the System.Data.SqlClient code, I see the same WindowsIdentity.GetCurrent() is called:
https://referencesource.microsoft.com/#system.data/system/data/providerbase/DbConnectionPoolIdentity.cs,81

This piece of logic in System.Data.SqlClient and Microsoft.Data.SqlClient is the same. There could be some extra steps for identity handled by SSMS and VS Code plugin to cause the WindowsIdentity.GetCurrent() return the expected user when impersonation happens. 

Thanks,
Karina

**Comment by cnayan** [comment]
@karinazhou Thank you for responding!

I also have noticed that my query was not returning data on remote SQL server - not an authentication issue :-)

But here is what I have and found:
- I don't have control over AD forest and trust relationships.
- The domains A & B may not have trust relationship established.. I don't know how it impacts, as I'm not expert in this area.
- LogonUser method did not work for me, except for "LOGON32_LOGON_NEW_CREDENTIALS" log on type. Thus, I also opted for log on provider as "LOGON32_PROVIDER_WINNT50". (To understand the relationship, check, [PInvoke page](http://pinvoke.net/default.aspx/advapi32/LogonUser.html), and [MS Docs page](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-logonusera)).
- [This answer](https://stackoverflow.com/a/11672293/315151) on Stackoverflow solved my problem completely. It helped me understand the missing pieces in the flow. I was mainly missing `DuplicateToken` call.
- I enhanced code in that answer. It has been published here: [User Impersonator](https://gist.github.com/cnayan/4a9f592b50437d14fae2949439074803)

I have the functionality working. Sorry for the trouble!
